### PR TITLE
Fix for alliance xp.

### PR DESCRIPTION
**Pull request by jwonesh**
Iterate over all available parties. Moved previous loop to inner loop (that is why you see a bunch of indents).

### CODE REVIEW

**Comment by jwonesh** [comment]
This is not ready, I see something I missed. Will fix tomorrow as it is late already.

**Comment by teschnei** [comment]
tabs -> spaces

**Comment by jwonesh** [comment]
Fixed. Github doesn't seem to be smart enough to ignore total whitespace changes (even if they are just spaces), so you can add ?w=1 to PR link to ignore whitespaces when diffing.

**Comment by jwonesh** [comment]
Also, I just had a thought that I need to make sure that MLevel is being calculated correctly. If we are in alliance, MLevel should be the highest person's level, yeah? I need to double check that is what it is doing when I get home form work. If it is only grabbing party's highest level, then we can have 2 alliances, one full of 75s and the other of low levels, where the 75 party can steamroll mobs and the low level party would get xp relative to their level.

**Comment by teschnei** [comment]
Yeah, but I don't want tabs in the repository :)
also, when using w=1 I can't comment on specific lines

I am pretty sure the main level is okay.  Whenever someone acts on a mob, it sets the level to their level if it is lower than their level.  So it should be alliance agnostic

**Comment by jwonesh** [comment]
Ok that is what I was thinking too.

I did realize that I will need to add a check if there is an alliance just before setting the PMember variable, as it will be null if I try to do so when no alliance exists. Didn't catch that when testing as I only made an alliance and saw that it worked. I will fix that when I get home from work.

File: src/map/utils/charutils.cpp [hunk 1]
```diff
@@ -3075,233 +3075,241 @@ namespace charutils
...
+			                {
+				                if (PMember->StatusEffectContainer->HasStatusEffect(EFFECT_SIGNET) && PMob->m_Element > 0 && dsprand::GetRandomNumber(100) < 20 &&
+					                PMember->loc.zone == PMob->loc.zone) // Need to move to SIGNET_CHANCE constant
+				                {
+					                PMember->PTreasurePool->AddItem(4095 + PMob->m_Element, PMob);
+				                }
+			                }
+			                if (PChar->PParty != nullptr && PChar->PParty->m_PAlliance != nullptr && PMob->m_Type == MOBTYPE_NORMAL)
+			                {
+				                if ((Pzone > 38 && Pzone < 43) || (Pzone > 133 && Pzone < 136) || (Pzone > 184 && Pzone < 189)) AddExperiencePoints(false, PMember, PMob, exp, 1, false);
```
Comment (xdemolish) [inline src/map/utils/charutils.cpp]: Pzone->getType() != ZONETYPE_DYNAMIS

**Comment by teschnei** [comment]
seems your tabs -> spaces missed a bunch of tabs still

**Comment by jwonesh** [comment]
Hmm I set visual studio to use 4 spaces instead of tab and redid all of the tabs. Let me take another look.

**Comment by teschnei** [comment]
CTRL+K, CTRL+F is auto-format selection (which should be good)
or CTRL+K, CTRL+D is auto-format entire document (which will be good if the document has been auto-formatted before, no guarantees on that one)

**Comment by jwonesh** [comment]
I reformatted per your suggestion @teschnei

**Comment by teschnei** [comment]
Great! I will get to the actual review tonight then

**Comment by jwonesh** [comment]
I would still recommend (after verifying that I have not introduced tabs into the repo) using the ignore-whitespace flag when viewing the diff, as Git still seems to think a lot more has changed than what really has.

**Comment by teschnei** [comment]
Yep, I saw the actual changes :)

File: src/map/utils/charutils.cpp [hunk 2]
```diff
@@ -3075,30 +3075,37 @@ namespace charutils
...
-                {
-                    exp = 0;
-                    monsterbonus = 1.0f;
-                    chainactive = false;
-                    CCharEntity* PMember = (CCharEntity*)PChar->PParty->members[i];
-                    if (PMember->getZone() == PMob->getZone())
-                    {
-                        if (map_config.exp_party_gap_penalties == 1)
+                uint8 numAllianceParties = 1;
+                if (PChar->PParty->m_PAlliance) numAllianceParties = PChar->PParty->m_PAlliance->partyList.size();
```
Comment (teschnei) [inline src/map/utils/charutils.cpp]: is it possible to use PChar->ForAlliance and use a lambda?  I want to hide the partylist/memberlist eventually so if you could get it to use that instead (if possible) I'd like that

**Comment by jwonesh** [comment]
@teshnei I'll see what I can do tonight

**Comment by jwonesh** [comment]
@teschnei I have rewritten to use lambda but I now get a compilation error which I cannot seem to fix, no matter how many stack overflow posts I read. Admittedly, my c++ skills are very basic (as I work with Java, normally), and my functional programming skills are even less developed. Here is the error that I get:

Error	C2893	Failed to specialize function template 'unknown-type std::invoke(_Callable &&,_Types &&...)'	DSGame-server	c:\microsoft visual studio 14.0\vc\include\type_traits	1494

It seems to have something do with using parameterized classes inside of the Lambda, but I cannot figure anything further out.

**Comment by teschnei** [comment]
commit your code and I can look at it

**Comment by jwonesh** [comment]
The solution became so clear once I took a step back and committed the code lol. I am recommitting the code for you to look at.

**Comment by teschnei** [comment]
I added the alliance penalty values and optimized it a tad, thanks for contributing :)

**Comment by jwonesh** [comment]
No problem, sorry my changes couldn't be of more use.

**Comment by teschnei** [comment]
your changes were fine, all I did was remove some duplicated code because
of them :)

**Pull request merged**